### PR TITLE
[RLlib] Prioritized version of an episode-based replay buffer.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

Changing to the new `EnvRunner API` for sampling and therein to focus on episodes instead of batches, we need to have replay buffers that can handle episodes. The `EpisodeReplayBuffer` was a simple replay buffer to deal with sequences, but offers no prioritized replay. This PR wants to change that by adding a prioritized replay buffer based on storing episodes and replay the transitions within in a prioritized way.

## Related issue number

#42369 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/utils/replay_buffers/utils.py [hunk 1]
```diff
@@ -21,6 +23,38 @@
 logger = logging.getLogger(__name__)
 
 
+def update_priorities_in_episode_replay_buffer(
+    replay_buffer: EpisodeReplayBuffer,
```
Comment (sven1977) [inline rllib/utils/replay_buffers/utils.py:27]: nit: PrioritizedEpisodeReplayBuffer ?

File: rllib/utils/replay_buffers/utils.py [hunk 2]
```diff
@@ -21,6 +23,38 @@
...
+    config: AlgorithmConfigDict,
```
Comment (sven1977) [inline rllib/utils/replay_buffers/utils.py:28]: Let's not use `AlgorithmConfigDict` anymore, only ever `AlgorithmConfig` objects.

File: rllib/utils/replay_buffers/episode_replay_buffer.py [hunk 3]
```diff
@@ -126,14 +126,16 @@ def add(self, episodes: Union[List["SingleAgentEpisode"], "SingleAgentEpisode"])
                 eps_idx = self.episode_id_to_index[eps.id_]
                 existing_eps = self.episodes[eps_idx - self._num_episodes_evicted]
                 old_len = len(existing_eps)
-                self._indices.extend([(eps_idx, old_len + i) for i in range(len(eps))])
+                self._indices.extend(
```
Comment (sven1977) [inline rllib/utils/replay_buffers/episode_replay_buffer.py:129]: Question: Doesn't this additional None lead to an error when we unpack/interpret `self._indices` somewhere else in this class?

File: rllib/utils/replay_buffers/prioritized_episode_replay_buffer.py [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+            capacity=capacity, batch_size_B=batch_size_B, batch_length_T=batch_length_T
+        )
+
+        assert alpha > 0
+        self._alpha = alpha
+
+        # Initialize segment trees for the priority weights. Note, b/c the trees are
+        # binary we need for them a capacity that is an exponential of 2.
+        # Double it to enable temporary buffer overflow (we need then free nodes in
+        # the trees).
```
Comment (sven1977) [inline rllib/utils/replay_buffers/prioritized_episode_replay_buffer.py:35]:
```suggestion
        # Initialize segment trees for the priority weights. Note that b/c the trees are
        # binary, we need a capacity that is an exponential of 2.
        # Double it to enable temporary buffer overflow (we then always have free nodes in
        # the trees).
```

File: rllib/utils/replay_buffers/prioritized_episode_replay_buffer.py [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+        buffer will be allowed to overflow in a temporary fashion, i.e. during
+        the next addition of samples to the buffer an attempt is made to fall below
+        capacity again.
+
+        The user is advised to select a large enough buffer with regard to the maximum
+        expected episode length.
+
+        Args:
+            episodes: A list of `SingleAgentEpisode`s that contain the episode data.
+            weight: A starting priority for the samples in `episodes`.
```
Comment (sven1977) [inline rllib/utils/replay_buffers/prioritized_episode_replay_buffer.py:72]: "samples" -> "timesteps"

File: rllib/utils/replay_buffers/prioritized_episode_replay_buffer.py [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+        """
+
+        # TODO (sven, simon): Eventually allow here an array?
+        if weight is None:
+            weight = self._max_priority
+
+        if isinstance(episodes, SingleAgentEpisode):
```
Comment (sven1977) [inline rllib/utils/replay_buffers/prioritized_episode_replay_buffer.py:79]: better:

from ray.rllib.utils import force_list

...

episodes = force_list(episodes)

File: rllib/utils/replay_buffers/episode_replay_buffer.py (hunk 3 above)
Comment (simonsays1980) [reply rllib/utils/replay_buffers/episode_replay_buffer.py]: It does - I might have accidentially added this there while it should have gone into the prioritized buffer. Good catch!

**Review by sven1977 — APPROVED** [review]
LGTM.

**Pull request merged**